include/seqan3/alignment/pairwise/execution/alignment_executor_two_way.hpp

```diff
@@ -0,0 +1,227 @@
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides seqan3::detail::alignment_executor_two_way.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <functional>
+#include <optional>
+#include <type_traits>
+
+#include <seqan3/alignment/pairwise/execution/alignment_range.hpp>
+#include <seqan3/alignment/pairwise/execution/execution_handler_sequential.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+#include <seqan3/std/ranges>
+
+namespace seqan3::detail
+{
+
+/*!\brief A two way executor for pairwise alignments.
+ * \ingroup execution
+ * \tparam align_instance_rng_t A bounded range over the alignment instances to compute.
+ * \tparam execution_handler_t  The execution handler managing the execution of the alignment instances.
+ */
+template <std::ranges::ViewableRange align_instance_rng_t,
+          typename execution_handler_t = execution_handler_sequential>
+class alignment_executor_two_way
+{
+    //!\brief Shortcut declaration for this class.
+    using this_t = alignment_executor_two_way<align_instance_rng_t, execution_handler_t>;
+
+    //!\brief Grants alignment_range access to the protected/private member of this class.
+    template <typename executor_t>
+    friend class alignment_range;
+
+    /*!\name Resource
+     * \{
+     */
+    //!\brief The iterator over the resource.
+    using resource_iterator = iterator_t<std::remove_reference_t<align_instance_rng_t>>;
+    //!\brief The sentinel over the resource.
+    using resource_sentinel = sentinel_t<std::remove_reference_t<align_instance_rng_t>>;
+    //!\brief The value type of the resource.
+    using resource_value_type     = std::remove_reference_t<decltype(*std::declval<resource_iterator>())>;
+    //!\}
+
+    /*!\name Buffer
+     * \{
+     */
+    //!\brief The result of invoking the alignment instance.
+    using buffer_value_type       = std::invoke_result_t<resource_value_type>;
+    //!\brief The internal buffer.
+    using buffer_type             = std::vector<buffer_value_type>;
+    //!\brief The pointer type of the buffer.
+    using buffer_pointer         = iterator_t<buffer_type>;
+    //!\}
+public:
+
+    //!\brief The result type of invoking the alignment instance.
+    using value_type = buffer_value_type;
+    //!\brief A reference to the alignment result.
+    using reference  = std::add_lvalue_reference_t<value_type>;
+    //!\brief The offset type for the buffer.
+    using off_type   = typename buffer_type::difference_type;
```

Comment:
why not call it `difference_type`?

Reply:
I probably change it back. I kind of see it as a stream, and wasn't sure if we might want to add seeking. So as if to skip alignment computations. But maybe that is over-engineered at the current point.